core/src/main/java/ai/timefold/solver/core/impl/solver/termination/DiminishedReturnsScoreRingBuffer.java

```diff
@@ -0,0 +1,224 @@
+            System.arraycopy(scoreRingBuffer, 0, newScoreRingBuffer, firstLength, secondLength);
+            readIndex = 0;
+            writeIndex = totalLength;
+        }
+        nanoTimeRingBuffer = newNanoTimeRingBuffer;
+        scoreRingBuffer = newScoreRingBuffer;
+    }
+
+    public void clear() {
+        readIndex = 0;
+        writeIndex = 0;
+        Arrays.fill(nanoTimeRingBuffer, 0);
+        Arrays.fill(scoreRingBuffer, null);
+    }
+
+    /**
+     * Returns the first element of the score ring buffer.
+     * Does not remove the element.
+     *
+     * @return the first element of the score ring buffer
+     */
+    public @NonNull Score_ peekFirst() {
+        return scoreRingBuffer[readIndex];
+    }
+
+    /**
+     * Adds a time-score pairing to the ring buffers, resizing the
+     * ring buffers if necessary.
+     *
+     * @param nanoTime the {@link System#nanoTime()} when the score was produced.
+     * @param score the score that was produced.
+     */
+    public void put(long nanoTime, @NonNull Score_ score) {
+        if (nanoTimeRingBuffer[writeIndex] != 0L) {
+            resize();
+        }
+        nanoTimeRingBuffer[writeIndex] = nanoTime;
+        scoreRingBuffer[writeIndex] = score;
+        writeIndex = (writeIndex + 1) % nanoTimeRingBuffer.length;
+    }
+
+    /**
+     * Removes count elements from both ring buffers,
+     * and returns the next element (which remains in the buffers).
+     *
+     * @param count the number of items to remove from both buffers.
+     * @return the first element in the score ring buffer after the count
+     *         elements were removed.
+     */
+    private @NonNull Score_ clearCountAndPeekNext(int count) {
+        if (readIndex + count < nanoTimeRingBuffer.length) {
+            Arrays.fill(nanoTimeRingBuffer, readIndex, readIndex + count, 0L);
+            Arrays.fill(scoreRingBuffer, readIndex, readIndex + count, null);
+            readIndex += count;
+        } else {
+            int remaining = count - (nanoTimeRingBuffer.length - readIndex);
+            Arrays.fill(nanoTimeRingBuffer, readIndex, nanoTimeRingBuffer.length, 0L);
+            Arrays.fill(scoreRingBuffer, readIndex, nanoTimeRingBuffer.length, null);
+            Arrays.fill(nanoTimeRingBuffer, 0, remaining, 0L);
+            Arrays.fill(scoreRingBuffer, 0, remaining, null);
+            readIndex = remaining;
+        }
+        return scoreRingBuffer[readIndex];
+    }
+
+    /**
+     * Returns the latest score prior or at the given time, and removes
+     * all time-score pairings prior to it.
+     *
+     * @param nanoTime the queried time in nanoseconds.
+     * @return the latest score prior to the given time.
+     */
+    public @NonNull Score_ pollLatestScoreBeforeTimeAndClearPrior(long nanoTime) {
+        if (readIndex == writeIndex && nanoTimeRingBuffer[writeIndex] == 0L) {
+            throw new IllegalStateException("Ring buffer is empty");
```

Comment:
Do you want to treat this as an exception? I see two options here:

- Either this is an exceptional state which must not happen, because the algorithm would not be correct. Then the message should start with "Impossible state: ", and I'm going to argue it should be thrown by the code calling this method, not by this method directly.
- Or this is a totally possible state, which we just want to signal to the calling code. In that case, this exception should be replaced by a return on `null` - see our previous conversation on exceptions and control flow.

Reply:
Exceptional state which must not happen; correctness assumes there at least one element in the buffer (which is inserted on start & whenever grace is reset).